src/Common/src/System/Globalization/IdnMapping.Unix.cs

```diff
@@ -67,5 +73,30 @@ private uint Flags
                 return (uint)flags;
             }
         }
+
+        /// <summary>
+        /// ICU doesn't check for invalid characters unless the STD3 rules option
+        /// is enabled.
+        ///
+        /// To match Windows behavior, we walk the string ourselves looking for these
+        /// bad characters so we can continue to throw ArgumentException in these cases. 
+        /// </summary>
+        private static void CheckInvalidIdnCharacters(string s, uint flags, string paramName)
+        {
+            if ((flags & Interop.GlobalizationNative.UseStd3AsciiRules) == 0)
+            {
+                for (int i = 0; i < s.Length; i++)
```

Comment:
this looks ok to me. if we think want to optimize the loop we can do something like the following.

            int length = s.Length;
            int index = 0;
            bool result = false;

            while (length > 0)
            {
                int len = Math.Min(length, 10);
                switch (len)
                {
                    case 10: c = s[index]; result |= (c <= 0x1F || c == 0x7F); index++;
                    case 9:  c = s[index]; result |= (c <= 0x1F || c == 0x7F); index++;
                    case 8:  c = s[index]; result |= (c <= 0x1F || c == 0x7F); index++;
                    case 7:  c = s[index]; result |= (c <= 0x1F || c == 0x7F); index++;
                    case 6:  c = s[index]; result |= (c <= 0x1F || c == 0x7F); index++;
                    case 5:  c = s[index]; result |= (c <= 0x1F || c == 0x7F); index++;
                    case 4:  c = s[index]; result |= (c <= 0x1F || c == 0x7F); index++;
                    case 3:  c = s[index]; result |= (c <= 0x1F || c == 0x7F); index++;
                    case 2:  c = s[index]; result |= (c <= 0x1F || c == 0x7F); index++;
                    case 1:  c = s[index]; result |= (c <= 0x1F || c == 0x7F); index++;
                }

                if (result)
                {
                    throw new ArgumentException(SR.Argument_IdnIllegalName, paramName); 
                }

                length -= len;
            }